File: semester-2/lab-7/C#/ShortList.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

namespace lab_7
{
    public class ShortList : IEnumerable<short>
    {

        private Node _head;
        private int _count;

        public ShortList()
        {
            _head = null;
            _count = 0;
        }

        public void Add(short data)
        {
            Node node = new Node(data);
            node.Next = _head;
            _head = node;
            _count++;
        }

        public bool Remove(short data)
        {
            Node current = _head;
            Node previous = null;
            if (Contains(data))
            {
                while (current != null)
                {
                    if (current.Data.Equals(data))
                    {
                        if (previous != null)
                        {
                            previous.Next = current.Next;
                        }
                        else
                        {
                            _head = _head.Next;
                        }

                        _count--;
                        return true;
                    }

                    previous = current;
                    current = current.Next;
                }
            }

            return false;
        }

        public bool Contains(short data)
        {
            return this.Any(value => value == data);
        }

        public double Average()
        {
            double average = 0;
            foreach (short data in this)
            {
                average += data;
            }

            return average / Count;
        }

        public int MultiplesOfThree()
        {
            int mot = 0;
            foreach (short data in this)
            {
                if (data % 3 == 0) mot++;
            }

            return mot;
        }

        public bool RemoveAboveAverage()
        {
            double average = Average();
            int removed = 0;
            foreach (short data in this)
            {
                if (data > average)
                {
                    Remove(data);
                    removed++;
                }
            }

            return Convert.ToBoolean(removed);
        }

        public int Count => _count;

        public void Display()
        {
            int i = 0;
            foreach (short data in this)
            {
                Console.Write(data + (i != Count - 1 ? " -> " : ""));
                i++;
            }

            Console.WriteLine();
        }

        public IEnumerator<short> GetEnumerator()
        {
            Node current = _head;
            while (current != null)
            {
                yield return current.Data;
                current = current.Next;
            }
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }
    }
}

File: semester-2/lab-7/C++/ShortList.cpp
#include <iostream>
#include "ShortList.h"

void ShortList::insert(short value) {
    Node *temp = new Node;
    temp->data = value;
    temp->next = head;
    head = temp;
}


void ShortList::remove(short data) {
    Node *current = head;
    Node *previous = nullptr;
    if(this->contains(data)) {
        while (current != nullptr) {
            if (current->data == data) {
                if (previous != nullptr) {
                    previous->next = current->next;
                } else {
                    head = head->next;
                }
            }
            previous = current;
            current = current->next;
        }
    }
}

void ShortList::display() {
    Node *current = head;
    while (current != nullptr) {
        cout << current->data << "\t";
        current = current->next;
    }
    cout << endl;
}


double ShortList::average() {
    Node *current = head;
    double average = 0;
    while (current != nullptr) {
        average += current->data;
        current = current->next;
    }
    return average / this->length();
}

int ShortList::multiplesOfThree() {
    Node *current = head;
    int mot = 0;
    while (current != nullptr) {
        if (current->data % 3 == 0) mot++;
        current = current->next;
    }
    return mot;
}

void ShortList::removeAboveAverage() {
    Node *current = head;
    double average = this->average();
    while (current != nullptr) {
        if (current->data > average) this->remove(current->data);
        current = current->next;
    }
}

int ShortList::length() {
    Node *current = head;
    int length = 0;
    while (current != nullptr) {
        length++;
        current = current->next;
    }
    return length;
}

bool ShortList::contains(short data) {
    Node *current = head;
    while (current != nullptr) {
        if (current->data == data) return true;
        current = current->next;
    }
    return false;
}
